Reporting Revenue is now more performant and scaleable

kjohnson opened the pull request:
<!-- Indicate the issue(s) resolved by this PR. -->

Resolves #5475

## Description

<!-- Summarize the related issue, explain HOW this PR solves the problem, and WHY you made the choices you made. -->

This change improves the performance and scalability of the reporting revenue endpoints by implementing the optimized `Reports\PaymentsQuery` to replace the N+1 Query resulting from the `Give_Payments_Query`.

Further, the Reporting Endpoints have been extracted to a new Domain, `src/Reports`, as a self contained feature with a dedicated Service Provider.

## Visuals

<!-- Include screenshots or video to better communicate your changes. -->

The below results can be compared to the original results posted under #5475.

**Note**: The `query_count` column includes the base request queries and indicates the performance of the entire request (not just the endpoint processing).

### 100 Donations

<pre>
+---------------------------------------+------------------+-------------+
| route                                 | request_time     | query_count |
+---------------------------------------+------------------+-------------+
| /give-api/v2/reports/average-donation | 1.1944000720978  | 31          |
| /give-api/v2/reports/total-refunds    | 1.0796999931335  | 31          |
| /give-api/v2/reports/top-donors       | 4.4131000041962  | 16          |
| /give-api/v2/reports/income           | 0.68830013275146 | 28          |
| /give-api/v2/reports/total-income     | 1.123899936676   | 31          |
| /give-api/v2/reports/income-breakdown | 1.7913999557495  | 68          |
| /give-api/v2/reports/payment-statuses | 0.16120004653931 | 15          |
| /give-api/v2/reports/total-donors     | 1.0538001060486  | 31          |
| /give-api/v2/reports/payment-methods  | 0.33550000190735 | 15          |
| /give-api/v2/reports/form-performance | 0.41580009460449 | 15          |
| /give-api/v2/reports/recent-donations | 3.3474998474121  | 15          |
+---------------------------------------+------------------+-------------+
</pre>

### 1,000 Donations

<pre>
+---------------------------------------+------------------+-------------+
| route                                 | request_time     | query_count |
+---------------------------------------+------------------+-------------+
| /give-api/v2/reports/total-refunds    | 1.6306002140045  | 31          |
| /give-api/v2/reports/average-donation | 1.769700050354   | 31          |
| /give-api/v2/reports/income           | 1.0776000022888  | 28          |
| /give-api/v2/reports/total-income     | 1.7267999649048  | 31          |
| /give-api/v2/reports/payment-statuses | 0.64779996871948 | 15          |
| /give-api/v2/reports/payment-methods  | 0.45939993858337 | 15          |
| /give-api/v2/reports/total-donors     | 1.672600030899   | 31          |
| /give-api/v2/reports/form-performance | 0.6746997833252  | 15          |
| /give-api/v2/reports/income-breakdown | 2.86510014534    | 68          |
| /give-api/v2/reports/recent-donations | 3.3203999996185  | 15          |
| /give-api/v2/reports/top-donors       | 4.1054000854492  | 16          |
+---------------------------------------+------------------+-------------+
</pre>


| Develop | Branch |
| --- | --- |
| ![Screen Shot 2020-12-04 at 14 18 24](https://user-images.githubusercontent.com/10858303/101205600-d9348400-363b-11eb-8850-82f196d90e62.png) | ![Screen Shot 2020-12-04 at 14 20 39](https://user-images.githubusercontent.com/10858303/101205628-e6517300-363b-11eb-980f-45c1226e4479.png) |

## Pre-review Checklist

<!-- Complete tasks prior to requesting a review. Add to this list, but do not remove the base items. -->

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions

<!-- Help others test your PR as efficiently as possible. -->

- Seed the database with a large set of donations.
- Load the reports for All Time on `develop`.
- Load the reports for All Time on `enhancement/improved-revenue-reporting`.
- Note the time difference in loading and the accuracy of the calculations.

ravinderk reviewed (changes requested):
@kjohnson Relocation file can cause of fatal error.

I found one conflict with Funds addon in which we are using `src/API/Endpoints/Reports/Endpoint.php`.

Can we keep backward compatibility if rename and relocation files is necessary?

kjohnson commented:
@ravinderk where in `give-funds` are we using the reports endpoint class?

Do you mean `Give\API\RestRoute`?

ravinderk commented:
@kjohnson You are correct, I got this error:

```
[01-Dec-2020 10:20:58 UTC] PHP Fatal error:  Interface 'Give\API\RestRoute' not found in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give-funds/src/Routes/Endpoint.php on line 14
```

kjohnson commented:
@ravinderk resolved the `API/RestRoute` backwards compatibility by 5ce355da9

ravinderk reviewed (changes requested):
@kjohnson I think update report endpoints are not loading correctly. I am setting unexpected results on the `Report page`.

![image](https://user-images.githubusercontent.com/1784821/100975443-0877c200-3564-11eb-8ade-d77f9ed7cb58.png)

kjohnson commented:
@ravinderk Hmm... Do you have some test data seeded in the database?

Can you inspect the network calls and see if there are any errors?

Here is my side by side comparison of develop vs the changes in this branch. The only difference is that develop never finished loading the top donors section.

| Develop | Branch |
| --- | --- |
| ![Screen Shot 2020-12-04 at 14 18 24](https://user-images.githubusercontent.com/10858303/101205600-d9348400-363b-11eb-8850-82f196d90e62.png) | ![Screen Shot 2020-12-04 at 14 20 39](https://user-images.githubusercontent.com/10858303/101205628-e6517300-363b-11eb-980f-45c1226e4479.png) |

ravinderk commented:
@kjohnson I can see inconsistency in results. We catch these issues with unit tests but we did not have enough tests. Before merging this pr, we can write more tests to catch inconsistent results.

| Develop | Branch |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1784821/103529421-a801da80-4eab-11eb-8686-4213faf94bc5.png) | ![image](https://user-images.githubusercontent.com/1784821/103529595-f31bed80-4eab-11eb-8178-f2b3ac4e8b20.png) |

kjohnson commented:
@ravinderk are you seeing any inconsistencies other than the Top Donor widget?

In my comparison (using All Time), the Top Donor timed out on `develop`, so this widget I did not compare.

kjohnson commented:
@ravinderk I reverted the custom query that I wrote for Top Donors and instead rely upon the improvements to the underlying `getPayments()` improvements.

ravinderk commented:
@kjohnson Get rid of the custom query will not resolve the issue and the requested api response still will be different from the previous one. Under the hood `BaseEndpoint::getPayments` perform custom query.

I also notice that I was not able to see donor Gravatars under the `Top Donors` card.

| Current | Develop
|---------|---------|
|![image](https://user-images.githubusercontent.com/1784821/103640099-6f2e3800-4f75-11eb-93a4-465c115516d8.png)|![image](https://user-images.githubusercontent.com/1784821/103640202-96850500-4f75-11eb-8662-0000f7c50214.png)|

I am getting following PHP errors when load Reports page:

```
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 72
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:72
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 163
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. Give\Reports\Endpoints\TotalIncome->get_trend() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:119
[05-Jan-2021 11:04:21 UTC] PHP  17. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:163
[05-Jan-2021 11:04:21 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, float given in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php on line 171
[05-Jan-2021 11:04:21 UTC] PHP Stack trace:
[05-Jan-2021 11:04:21 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:21 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:21 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:21 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:21 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:21 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:21 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:21 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:21 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:21 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:21 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:21 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:21 UTC] PHP  13. Give\Reports\Endpoints\TotalIncome->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:21 UTC] PHP  14. Give\Reports\Endpoints\TotalIncome->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:21 UTC] PHP  15. Give\Reports\Endpoints\TotalIncome->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:32
[05-Jan-2021 11:04:21 UTC] PHP  16. Give\Reports\Endpoints\TotalIncome->get_trend() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:119
[05-Jan-2021 11:04:21 UTC] PHP  17. array_filter() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TotalIncome.php:171
[05-Jan-2021 11:04:28 UTC] PHP Notice:  Undefined property: stdClass::$payment_meta in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/RecentDonations.php on line 28
[05-Jan-2021 11:04:28 UTC] PHP Stack trace:
[05-Jan-2021 11:04:28 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:28 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:28 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:28 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:28 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:28 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:28 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:28 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:28 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:28 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:28 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:28 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:28 UTC] PHP  13. Give\Reports\Endpoints\RecentDonations->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:28 UTC] PHP  14. Give\Reports\Endpoints\RecentDonations->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:28 UTC] PHP Notice:  Undefined property: stdClass::$email in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TopDonors.php on line 46
[05-Jan-2021 11:04:28 UTC] PHP Stack trace:
[05-Jan-2021 11:04:28 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:28 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:28 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:28 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:28 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:28 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:28 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:28 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:28 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:28 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:28 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:28 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:28 UTC] PHP  13. Give\Reports\Endpoints\TopDonors->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:28 UTC] PHP  14. Give\Reports\Endpoints\TopDonors->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:28 UTC] PHP  15. Give\Reports\Endpoints\TopDonors->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TopDonors.php:21
[05-Jan-2021 11:04:28 UTC] PHP Notice:  Undefined property: stdClass::$email in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TopDonors.php on line 47
[05-Jan-2021 11:04:28 UTC] PHP Stack trace:
[05-Jan-2021 11:04:28 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[05-Jan-2021 11:04:28 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[05-Jan-2021 11:04:28 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[05-Jan-2021 11:04:28 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[05-Jan-2021 11:04:28 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[05-Jan-2021 11:04:28 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[05-Jan-2021 11:04:28 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[05-Jan-2021 11:04:28 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[05-Jan-2021 11:04:28 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[05-Jan-2021 11:04:28 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[05-Jan-2021 11:04:28 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[05-Jan-2021 11:04:28 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[05-Jan-2021 11:04:28 UTC] PHP  13. Give\Reports\Endpoints\TopDonors->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[05-Jan-2021 11:04:28 UTC] PHP  14. Give\Reports\Endpoints\TopDonors->getReport() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/BaseEndpoint.php:122
[05-Jan-2021 11:04:28 UTC] PHP  15. Give\Reports\Endpoints\TopDonors->get_data() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Reports/Endpoints/TopDonors.php:21
```

I am getting the above errors because`Give\Reports\PaymentsQuery` is not equivalent to `Give_Payments_Query`. We do not support most of the query param like `order`, `order by`, `form id` etc. That means we are performing different SQL querying in this pull request. For the same reason, I see a different result in `develop`.


@JasonTheAdams I am looking forward to your suggestion because I can see that few quick patches will not make this pull request ready to release because there are chances to miss minor detail under less unit tested for reporting.

kjohnson commented:
I've added the donor email, so gravatars should now show - I didn't see this earlier because on `develop` "Top Donors" wouldn't even load at scale.

Also, I added support for Order By, although every report endpoint orders by date DESC - so this could probably be refactored out. The date column is always used even though it is a parameter, but the order direction is fixed.

@ravinderk can you check again for inconsistencies? I've mapped all of the existing `$args` to the new query. The query will be different, but it has to be different for "All Time" donations to process under load.

Another note: With this change I am sometimes seeing a difference in the "Revenue for Period" chart - I think this is because of a change to the date ranges. On `develop`, the end date string is appended with `24:00:00`, which causes an error in the MySQL syntax. In this change, the end string is appended with `23:59:59`

ravinderk commented on src/Reports/Endpoints/TotalIncome.php:
```diff
@@ -144,8 +154,22 @@ public function get_trend( $start, $end, $income ) {
 
-		$prevIncome    = $this->get_income( $prevStart->format( 'Y-m-d' ), $prevEnd->format( 'Y-m-d' ) );
-		$currentIncome = $this->get_income( $start->format( 'Y-m-d' ), $end->format( 'Y-m-d' ) );
+		// $prevIncome    = $this->get_income( $prevStart->format( 'Y-m-d' ), $prevEnd->format( 'Y-m-d' ) );
```
@kjohnson Can explain how you are calculating trends with the same data? I can not see trends in `Total Revenue`

![image](https://user-images.githubusercontent.com/1784821/108051910-7450cd80-7071-11eb-9681-dcaf714d368f.png)

ravinderk commented on src/Reports/PaymentsQuery.php:
```diff
@@ -0,0 +1,151 @@
+                AND DonorLastName.meta_key = '_give_donor_last_name'
+			JOIN {$wpdb->prefix}give_donors as DonorEmail
+                ON DonorID.meta_value = DonorEmail.id
+			WHERE Donation.post_type = 'give_payment'
```
@kjohnson We are not querying renewals in this query. After discussion with @JasonTheAdams, I opened an issue to make renewal queriable with a simple query https://github.com/impress-org/give-recurring/issues/1030

ravinderk commented on src/Reports/ServiceProvider.php:
```diff
@@ -0,0 +1,51 @@
+	/**
+	 * Calls the route registrations within the WordPress REST API hook
+	 *
+	 * @since 2.8.0
```
@kjohnson Can we update all `@since` tags to `2.10.0`

ravinderk commented on src/Reports/PaymentsQuery.php:
```diff
@@ -0,0 +1,151 @@
+
+namespace Give\Reports;
+
+/**
```
@kjohnson `@since` tags are missing from most of the files? Can we add it?

ravinderk commented on src/Reports/Endpoints/BaseEndpoint.php:
```diff
@@ -354,6 +354,8 @@ private function getCachedPayments( $args ) {
 	/**
 	 * Get payment.
 	 *
+	 * @since 2.9.5 Replaced `Give_Payments_Query` with an optimized query.
```
@kjohnson Can we update the `@since` tag to `2.10.0`

ravinderk reviewed (changes requested):
@kjohnson I only see an issue with `Total Revenue Trends`. I also suggested updating the `@since` tag.

kjohnson commented:
Closing this for now - I'll revisit after scaffolding some tests for Reports.
